src/LauncherManager.vala

```diff
@@ -0,0 +1,221 @@
+/*
+ * SPDX-License-Identifier: GPL-3.0
+ * SPDX-FileCopyrightText: 2023 elementary, Inc. (https://elementary.io)
+ */
+
+ public class Dock.LauncherManager : Gtk.Fixed {
+    public const string ACTION_GROUP_PREFIX = "app-actions";
+    public const string ACTION_PREFIX = ACTION_GROUP_PREFIX + ".";
+    // First %s is the app id second %s the action name
+    public const string LAUNCHER_ACTION_TEMPLATE = "%s.%s";
+    // %s is the app id
+    public const string LAUNCHER_PINNED_ACTION_TEMPLATE = "%s-pinned";
+
+    private static Settings settings;
+
+    private static GLib.Once<LauncherManager> instance;
+    public static unowned LauncherManager get_default () {
+        return instance.once (() => { return new LauncherManager (); });
+    }
+
+    private SimpleActionGroup action_group;
+    private List<Launcher> launchers; //Only used to keep track of launcher indices
+    private Dock.DesktopIntegration desktop_integration;
+    private GLib.HashTable<unowned string, Dock.Launcher> app_to_launcher;
+
+    static construct {
+        settings = new Settings ("io.elementary.dock");
+    }
+
+    construct {
```

Comment:
I think this widget also needs `overflow = VISIBLE` to not clip launcher icons when they bounce

Reply:
Done :)